### ISSUE TITLE
[Question] loss frequently drops to 0.2 while training second-stage of llava-v1.5-13b

### ISSUE THREAD

**Issue by Maxlinn** [haotian-liu/LLaVA#579]
### Question

hi liu and the team, thanks for bringing about llava!

recently i'm trying to reproduce the seond-stage(instruct fine-tuning) of llava-v1.5-13b. i followed the guidance of [LLaVA#visual-instruction-tuning](https://github.com/haotian-liu/LLaVA#visual-instruction-tuning) to prepare data and run. the pretrained first-stage mm_projector was obtained from [liuhaotian/llava-v1.5-mlp2x-336px-pretrain-vicuna-13b-v1.5](https://huggingface.co/liuhaotian/llava-v1.5-mlp2x-336px-pretrain-vicuna-13b-v1.5). 

i trained on a single node with 8x80G vram. the start script is identical to [scripts/v1_5/finetune.sh](https://github.com/haotian-liu/LLaVA/blob/main/scripts/v1_5/finetune.sh), however after training somewhile, the loss diagram seems strange, as it frequently drops to about 0.2 and recover to around 0.8.

![image](https://github.com/haotian-liu/LLaVA/assets/43326876/0dc00836-cfb8-498a-9885-f98f6704d03f)

at first, i was skeptical of ocr-vqa data since it needs crawling. so i removed all 80000 ocr-vqa converstations from llava_v1_5_mix665k.json. however, the loss still drops often.

![image](https://github.com/haotian-liu/LLaVA/assets/43326876/9920bea2-c6c8-4966-b11a-e2dca5ed0f9b)

is this normal? could you please give me some hint? much thanks!

**Comment by Maxlinn** [comment]
i did some ablation on data and found it occurs when image-text instructions **mixed with pure text instructions**, with `--group_by_modality_length True`.

i did exp on training with pure coco, gqa, textvqa with `--group_by_modality_length False`, the cuves has no down spikes.

![image](https://github.com/haotian-liu/LLaVA/assets/43326876/963c9395-84a8-44fe-9fd5-1588b8fb1de4)

it is understandable, but i still want to confirm that down spike occurs when llava-v1.5-13b trained officially?

**Comment by xmy0916** [comment]
i have trained my model for 5 epoch, i discovered other interesting phenomena, the loss seems to decrease significantly at the beginning of each epoch, and then remains basically the same throughout the epoch.
![image](https://github.com/haotian-liu/LLaVA/assets/43675899/f08f66ee-31f7-4d37-b2a1-fce90d7a1f0c)

**Comment by Maxlinn** [comment]
> i have trained my model for 5 epoch, i discovered other interesting phenomena, the loss seems to decrease significantly at the beginning of each epoch, and then remains basically the same throughout the epoch. ![image](https://user-images.githubusercontent.com/43675899/275329695-f08f66ee-31f7-4d37-b2a1-fce90d7a1f0c.png)

much thanks for sharing the diagram! 

based on my previous experiences on VLM, the fast-then-flat tendency on loss surely exists(despite mixture of textual data causing down spikes). the loss is dropping slowly afterwards but the training is still important for better model evalutaion, the loss does not necessaily reflects performance.

by the way, i wonder if you ever come across that when training on 8 gpus, the first gpu uses up vram but do not have any usage?  when i was doing instruct fine-tuning on llava it appears so, very strange.

**Comment by xmy0916** [comment]
@Maxlinn the first gpu uses up vram but do not have any usage? when i was doing instruct fine-tuning on llava it appears so, very strange。sry，这句话我没看懂啥意思，要不in chinese？hhhh，你的意思是第一张gpu显存打满了，但是利用率为0？

**Comment by Maxlinn** [comment]
> @Maxlinn the first gpu uses up vram but do not have any usage? when i was doing instruct fine-tuning on llava it appears so, very strange。sry，这句话我没看懂啥意思，要不in chinese？hhhh，你的意思是第一张gpu显存打满了，但是利用率为0？

是的，八张卡训练，第一张卡显存用满了但是利用率一直是0，后面7张卡显存和占用率都是满的，用的是[haotian-liu/LLaVA#visual-instruction-tuning](https://github.com/haotian-liu/LLaVA#visual-instruction-tuning)的脚本。

**Comment by haotian-liu** [comment]
Hi, these losses are normal. The spikes are mainly due to the ShareGPT data used as you observed in your ablation, as our base model Vicuna has already been trained on these samples (probably for 3 epochs). We use these samples as a sort of regularization to prevent forgetting of the LLM knowledge/capabilities. It is perhaps a good phenomenon for us to see the spikes (to low loss). Also, the sudden drop in losses between epochs is observed in both LLM training and LMM training.

Below I provide the loss curves for both stage training of LLaVA-v1.5-13B for your reference.

![image](https://github.com/haotian-liu/LLaVA/assets/6631389/2753dfb3-b409-4f2e-9e21-950b529ed991)

![image](https://github.com/haotian-liu/LLaVA/assets/6631389/c3bc77ab-1e16-4058-820b-de00214acdc5)

**Comment by Maxlinn** [comment]
thanks for sharing the diagram! i observed similiar loss values and spikes so i guess nothing wrong.

**Comment by xmy0916** [comment]
> > @Maxlinn the first gpu uses up vram but do not have any usage? when i was doing instruct fine-tuning on llava it appears so, very strange。sry，这句话我没看懂啥意思，要不in chinese？hhhh，你的意思是第一张gpu显存打满了，但是利用率为0？
> 
> 是的，八张卡训练，第一张卡显存用满了但是利用率一直是0，后面7张卡显存和占用率都是满的，用的是[haotian-liu/LLaVA#visual-instruction-tuning](https://github.com/haotian-liu/LLaVA#visual-instruction-tuning)的脚本。

这个我没太注意，后面再训练时候观察下

**Comment by Maxlinn** [comment]
> > > @Maxlinn the first gpu uses up vram but do not have any usage? when i was doing instruct fine-tuning on llava it appears so, very strange。sry，这句话我没看懂啥意思，要不in chinese？hhhh，你的意思是第一张gpu显存打满了，但是利用率为0？
> > 
> > 
> > 是的，八张卡训练，第一张卡显存用满了但是利用率一直是0，后面7张卡显存和占用率都是满的，用的是[haotian-liu/LLaVA#visual-instruction-tuning](https://github.com/haotian-liu/LLaVA#visual-instruction-tuning)的脚本。
> 
> 这个我没太注意，后面再训练时候观察下

最后发现是我机器显示的问题，与代码无关，感谢讨论！

**Comment by July-zh** [comment]
哈喽，我用一台8卡A100（40G）训练的loss曲线跟80G的有些出入，大家感觉这个正常吗？
配置是这样的：
![image](https://github.com/haotian-liu/LLaVA/assets/39486987/2ffa5fce-2e6f-4ee1-b95a-99253989688b)

![image](https://github.com/haotian-liu/LLaVA/assets/39486987/9798026f-0189-405e-b496-0043fbe5b659)

**Comment by haotian-liu** [comment]
Hi @July-zh 

This theoretically should not cause issues, as this is literally mixing text-only data and image-only data within gradient accumulation steps. This commit (f16049a3d35e281a4d88b7a2cb0a7f36a0c677ef) makes the gradient accumulation behavior consistent with large batch training. Either case, your training should be normal.

Also, (sorry for the delay), LoRA training checkpoints will be released in next few days (very likely tomorrow).

**Comment by linhaojia13** [comment]
> Also, the sudden drop in losses between epochs is observed in both LLM training and LMM training.

@haotian-liu However, LLaVA is trained for only 1 epoch. Is this because training for more than 1 epoch leads to overfitting, even though it might result in a lower training loss?

**Comment by bpwl0121** [comment]
> apabilities. It is perhaps a good phenomenon for us to see the spikes (to low loss). Also, the sudden drop in losses between epochs is observed in both LLM tr

hi haotian, thanks for you detailed reply. 
did you do any ablation study about the pure text dataset do prevent the forgetting of llava？
